### ISSUE TITLE
Incorrect Screen Reader focus on transition

### ISSUE THREAD

**Issue by niksawtschuk** [react-navigation/react-navigation#7056]
## Context
While auditing our application using VoiceOver on iOS, I noticed that screen transitions (pushing on a stack as well as presenting modally) don't yield behaviour I would expect from a native iOS application using UINavigation and UIViewControllers.

At minimum, I would expect that navigating to a new screen (either by pushing one on the stack, presenting one modally, or dismissing a screen presented in either way) would cause the the VoiceOver cursor to focus on the element in the top-left of the screen. Ideally, you would want to focus on the screen's title, or an appropriate heading/summary which would give enough context about what the user has transitioned to.

## Problem
Unfortunately, the behaviour I'm seeing is that the cursor is (for the most part) placed at the element in the same approximate location that as where it was on the previous screen. This is unexpected behaviour and leads to a sub-optimal experience for Screen Reader users.

## Solutions
Some possible solutions that I can see:
1. Listen for screen appearance inside [Header](https://github.com/react-navigation/stack/blob/master/src/views/Header/Header.tsx) and use [AccessibiltyInfo](https://facebook.github.io/react-native/docs/accessibilityinfo) to set the focus on the desired view.
2. Always pass a `headerTitle` component instead of a string when setting a screen's `navigationOptions`. Then, on the app, [listen for screen appearance](https://reactnavigation.org/docs/en/function-after-focusing-screen.html) and set the focus to the component you passed as `headerTitle`.

As a consumer, obviously I would prefer solution 1, I also think it would be great to provide this behaviour to all users of this library (perhaps behind a switch, at first?).

It's also possible that [using the native navigation components](https://reactnavigation.org/blog/2019/10/17/react-navigation-native.html) will resolve this, but I haven't tried.

## Notes
Package Versions:
```
"react-native": "0.61.4",
"react-navigation": "4.0.10",
"react-navigation-drawer": "2.3.3",
"react-navigation-stack": "1.10.3",
"react-navigation-tabs": "2.5.6",
```

**Comment by satya164** [comment]
This should be fixed in the latest version.

**Comment by howadev** [comment]
> This should be fixed in the latest version.

Still can reproduce with react-navigation@4.2.1

**Comment by satya164** [comment]
You need to update `react-navigation-stack` to `2.2.1`

**Comment by howadev** [comment]
> You need to update `react-navigation-stack` to `2.2.1`

I just update `react-navigation-stack` from `2.2.0` to `2.2.1` but still can reproduce the issue. Just curious which commit do you think fix the issue?

**Comment by satya164** [comment]
Can you provide a repro/video?

**Comment by howadev** [comment]
> Can you provide a repro/video?

demo repo and video: https://github.com/howadev/ReactNavigationVoiceOverDemo

As you can see, voice over doesn't focus to header title as expected after pushing to a new screen.

**Comment by renaudAmsellem** [comment]
Hey - thanks for the lib.

I can reproduce this behavior in the v5.

Package Versions:
```
"react-native": "0.61.3,
"react-navigation": "5.1.6,
"react-navigation-stack": "5.2.13",
```

@lissau i'd love if you could show a working example of the workaround

**Comment by vdmtrv** [comment]
This seems to be related to react-navigation rendering the header component **AFTER** all of the content, hence why you might hear the "end of content" screen reader sound when it thinks it has reached the end and loops over the items on the screen.

**Comment by summerjoyyoung** [comment]
Here's my work around solution. It worked in my use case. Hope it helps others!
`const Nav = createStackNavigator()
 
export default function MyNavigator() {
    // First you will want to create a ref for the view that you want to focus to
    const viewRef = useRef(null)
 
    return (
        <Nav.Navigator>
            <Nav.Screen
 
                ...[whatever options and screens you want to set up here]
 
                // Second set the view's ref prop is set to the viewRef constant above
                options={{
                    headerShown: true,
                    title: '',
                    headerLeft: () => (
                        <View
                            accessible={true}
                            ref={viewRef}
                            accessibilityRole="image">
                            <HumanaLogo fill={LIGHT_GREEN} />
                        </View>
                    ),
                }}
                // Lastly, set the accessibility focus to the viewRef in the transitionEnd listener
                listeners={{
                    transitionEnd: () => {
                        if (viewRef && viewRef.current) {
                            const reactTag = findNodeHandle(viewRef.current)
                            if (reactTag) {
                                AccessibilityInfo.setAccessibilityFocus(reactTag)
                            }
                        }
                    },
                }}
            />
        </Nav.Navigator>
    )
}`

**Comment by HermanNygaard** [comment]
Any updates on this?

**Comment by thibmaek** [comment]
The solution above works, but gets really nasty if you have a lot of child components. You need to forward the ref to the deepest element in order to let it focus on that element using RN's Accessibility API

```tsx
const Heading3 = React.forwardRef(({ children }), ref) => <Text ref={ref}>{children}</Text>;

const Header = React.forwardRef(({ title }), ref) => <View><Heading3 ref={ref}>{title}</Heading3></View>;

const NavHeader = React.forwardRef(({ title }), ref) => <><BackButton /><Header title={title}></>;

const RootNav = () => {
  const ref = useRef();

  return (
    <Stack.Navigator headerMode="screen">
      <Stack.Screen
        component={DetailView}
        options={{
          header: ({ navigation, scene, insets }) => (
            <NavHeader ref={focusRef} title="Detail" />
          ),
        }}
        listeners={{
          transitionEnd: () => {
            if (Platform.OS === 'ios') {
              if (ref.current) {
                const focusPoint = findNodeHandle(ref.current);
                if (focusPoint) {
                  AccessibilityInfo.setAccessibilityFocus(focusPoint);
                }
             }
           }
          },
        }}
      />
    </Stack.Navigator>
}
```

**Comment by ZeiRashed** [comment]
@satya164 I am still reproing this issue RN stack version 5.14.9. Do you know if there are plans to fix this? If so, what would the timeline be?

**Comment by efibe** [comment]
For those waiting for an elegant workaround, there is this one: 

https://dev.to/inthepocket/fixing-react-navigation-ios-screen-reader-focus-7op

**Comment by gregdburns** [comment]
For anyone who has struggled getting this to work on both Android and iOS, I've created a hook that is working in our app for both platforms.  Note the setTimeOut, which is key to getting it to work with the useFocusEffect hook.

Here's the hook:

```
import React, { useRef } from 'react'
import { AccessibilityInfo, findNodeHandle } from 'react-native'
import { useFocusEffect } from '@react-navigation/native'

/**
 * This hook should be used to programmatically set focus on a component when the screen is loaded.
 * It should be used as a last resort.  Ideally focus would be managed with correct order of components in the tree and proper use of the "accessible" prop.
 *
 * Components that require setting programmatic focus on events (ex: a toast a result of an action) should use AccessibilityInfo.setAccessibilityFocus directly.
 *
 * @returns {Object} A ref that should be assigned to the "ref" property of the component that should be focused on the screen.
 */
const useAccessibilityFocus = () => {
  const focusRef = useRef()
  useFocusEffect(
    React.useCallback(() => {
      if (focusRef.current) {
        const reactTag = findNodeHandle(focusRef.current)
        if (reactTag) {
          setTimeout(() => { // Due to screen lifecycle, we need to put a minor delay, or React Nav (or other) components may steal focus.
            AccessibilityInfo.setAccessibilityFocus(reactTag)
          }, 200)
        }
      }
    }, [focusRef.current])
  )

  return { focusRef }
}

export default useAccessibilityFocus
```
And here's an example implementation:

```
...functional component logic
const { focusRef } = useAccessibilityFocus()
...
return (
...
<Component ref={focusRef} />
)
```

**Comment by gregdburns** [comment]
An (unfortunate) addition to the workarounds proposed above.  There can still be inconsistent results and the hook can, at times, conflict with user interactions or other components that are trying to get focus.  While the workaround may be better than nothing, the ideal solution would be to have this fixed in the react-nav library.

**Comment by pandu-supriyono** [comment]
[Reproducible with this repo](https://github.com/pandu-supriyono/focus-hanging-on-tab-repro)

```
"@react-navigation/bottom-tabs": "^6.5.7",
"@react-navigation/native": "^6.1.6",
"@react-navigation/stack": "^6.3.16",
"react": "18.2.0",
"react-native": "0.71.5",
"react-native-gesture-handler": "^2.9.0",
"react-native-safe-area-context": "^4.5.0",
"react-native-screens": "^3.20.0"
```

I'm curious if there is any discourse, not necessarily related to this lib, with regards to expected user experience when it comes to navigating tabs. I see many other apps that do not have the desired behaviour as described in this issue, but at the very least they immediately announce the currently selected tab after the user has navigated to it.

If we are willing to talk in terms of WCAG, I interpret the lack of this behaviour as [a violation of 2.4.3: Focus Order](https://www.w3.org/TR/WCAG21/#focus-order). 

I hope we can shed more light on this issue and its implications in practice.

**Comment by mariushorvat** [comment]
I'm also experience these issue. Is there a plan to get this fixed? Any walkaround? Thanks!

**Comment by msftedad** [comment]
Hi @pandu-supriyono, Any update on this issue?

**Comment by msftedad** [comment]
Hi @pandu-supriyono, Any update on this issue?

**Comment by msftedad** [comment]
Hi @pandu-supriyono, Any update on this issue?

**Comment by wdospinal** [comment]
I'm also seeing this error with 
"react-native": "0.71.13",

**Comment by pandu-supriyono** [comment]
@msftedad unfortunately, none at the moment. I'll get back as soon as I learn something new.

**Comment by juaal12** [comment]
Hi guys!

Any plans to fix this issue?

In July 2025, WCAG 2.1 will be a crucial aspect of the European Accessibility Act (EAA). The EAA, which enters into force on June 28, 2025, mandates that many products and services, including websites and mobile apps, comply with accessibility requirements for people with disabilities.

With this on going issue, it is impossible to reach the accessibility goal at all :(

**Comment by msftedad** [comment]
Hi team, Any update on this issue?

**Comment by izyspania** [comment]
Im interested in a solution too .

**Comment by chramos** [comment]
lol, 5 years and we still have this issue... congratulations!
who needs accessibility!

**Comment by msftedad** [comment]
Hi @pandu-supriyono ,
Any update on this issue?

**Comment by ArturKalach** [comment]
Hello there,

Does anyone know if this issue is still reproducible?
If possible, could someone share a repo to help reproduce it?

Just tested it, and it works pretty reliably for me. Did I miss something?
```
    "@react-navigation/bottom-tabs": "^7.2.0",
    "@react-navigation/native": "^7.0.14",
    "@react-navigation/native-stack": "^7.2.0",
    "react": "18.3.1",
    "react-native": "0.76.5",
    "react-native-screens": "^4.5.0"
```

_____

I was looking for another issue and came across this one.
_____

@msftedad , @chramos 
Could you please share your experience, if possible?

**Comment by williamsaar** [comment]
@ArturKalach We're experiencing the same issue. Don't have the possibility to provide a repro at the current time, but here are our related dependencies.

```
    "@react-navigation/bottom-tabs": "^7.9.0",
    "@react-navigation/native": "^7.1.26",
    "@react-navigation/stack": "^7.6.13",
    "react": "19.1.0",
    "react-native": "0.81.5",
    "react-native-safe-area-context": "~5.6.0",
    "react-native-screens": "~4.16.0",
```
